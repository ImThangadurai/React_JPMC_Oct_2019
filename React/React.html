<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">

    <script crossorigin src="https://unpkg.com/react@16/umd/react.development.js"></script>
    <script crossorigin src="https://unpkg.com/react-dom@16/umd/react-dom.development.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-core/5.8.38/browser.js"></script>
    <script src="test.js"></script>
    <title>React</title>
</head>
<body>
    <h2>React App</h2>

    <div id="app">

    </div>
    <div id="app1">

    </div>

    <script type="text/babel">
        
        //var element = React.createElement("h3", null, "This is a react element");
        var element = React.createElement("h3", null, 
                        React.createElement("p", null, "This is a react element"));
        ReactDOM.render(element, document.getElementById("app"));

        //ReactDOM.render(<h3><p>This is React JSX</p></h3>, document.getElementById("app1"));

        var MyComponent = function(){

            return (
                <h3><p>This is React JSX</p></h3>
            );

        }
        ReactDOM.render(<MyComponent/>, document.getElementById("app1"));
        
    </script>


</body>
</html>