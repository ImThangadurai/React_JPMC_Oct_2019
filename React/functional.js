var numbers = [1,2,3,4,5,6];

var result = numbers.filter(n => n > 4);
console.log(result);

result = numbers.map(n => n * n);
console.log(result);