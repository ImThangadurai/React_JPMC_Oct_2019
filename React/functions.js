
// implicit args  ==> this, arguments
function sum(x, y){
    console.log("in sum");
    console.log("this", this);
    console.log("arguments", arguments);
}

// function sum(){
//     console.log("in sum no args");
// }
// sum(2,3);
// sum();
// sum(2,3,4,5,6);

//Arraw functions ==>  no args  ==> this, arguments

// Callbacks and Async code
// Functional API's

var obj = {
    id: 1,
    print: function(){
        console.log("id: ", this.id);
        // setTimeout(function(){
        //     console.log("id: ", this.id);
        // }, 2000);
        setTimeout(()=>{
            console.log("id: ", this.id);
        }, 2000);
    }
} 

obj.print();
