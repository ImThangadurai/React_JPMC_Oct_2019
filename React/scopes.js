console.log("x: ", x);
var x = 10;
console.log("x: ", x);

var y;;
console.log("y: ", y);

//console.log("z: ", z);

//function statement
function foo()
{
    console.log("foo...");
    if(x < 100){
        var msg = "Hello";
    }
    console.log("Msg: ", msg );
}
foo();
//IIFE (Immediately Invoked Function Expression)==> block scope

//function expression
var bar = function(){

};

(function(){
    console.log("iife...");
})();


function iffe()
{
    console.log("foo...");
    if(x < 100){
        (function(){
            var msg = "Test iife";
            console.log("Msg in iife: ", msg );
        })()
    }
    
}
iffe();

//ES6 Block scopes==> let const

function blockscopes(){

    if(x < 100){
        let test = "ABC";
        console.log("test in b_scopes: ", test);
    }
    //console.log("test in b_scopes: ", test);
}
blockscopes();


console.log("app over");