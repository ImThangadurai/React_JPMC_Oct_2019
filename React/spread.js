var numbers = [1,2,3];

var copy = [...numbers];

console.log(numbers);
console.log(copy);

numbers.push(4);

console.log(numbers);
console.log(copy);

var names = ["a", "b", "c"];

var merge = [...numbers, ...names, 10, "tests"];
console.log(merge);

var user = {
    id: 10, name: "Anil"
}

var userCopy = {...user, loc: "Mumbai", name: "Joseph"};
user.id = 100;
console.log(user);
console.log(userCopy);

