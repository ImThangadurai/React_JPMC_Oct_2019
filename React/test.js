var obj = {
    id: 10,
    print: function(){
        console.log("Id: ", this.id);
    },
    somefn: function(msg){
        console.log("Id: ", this.id);
        console.log("Msg: ", msg);
    }

}
obj.print();

var emp = {
    id: 200
};

//emp.print();
var fn=obj.print.bind(emp);
fn();

var fn1 = obj.somefn.bind(emp);
fn1("this is a message");

var fn2 = obj.somefn.bind(emp, "Test msg");
fn2();





