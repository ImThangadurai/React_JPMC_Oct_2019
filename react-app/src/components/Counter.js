import React, {Component} from 'react';

//ES 6
class Counter extends Component{
 
    //ES 7
    state = {
        count: 0
    }

    //inpRef = null;

    constructor(props){
        super(props);

        //ES6
        // this.state = {
        //     count: 0
        // };

        if(this.props.value){
            this.state.count = parseInt(this.props.value);
        }
        this.inc = this.inc.bind(this);
    }

    inc(){
        console.log("Counter inc");
        this.setState({
            count: this.state.count + 1
        });
    }

    //ES7
    decr = () => {
        console.log("Counter decr");
        this.setState({
            count: this.state.count - 1
        });
    } 
    change = (evt) => {
        console.log("change", evt.target.value);
        this.setState({
            count: parseInt(evt.target.value)
        });
    }

    setValue = () => {
        if(this.inpRef.value){

            this.setState({
                count: parseInt(this.inpRef.value)
            });
        }
    }
    render(){
        return (
            <div>
                <h3>{this.props.title}</h3>
                <h5>Counter: {this.state.count}</h5>

                <div>
                    {/* Controlled Input */}
                    Count: <input type="number" 
                                placeholder="Count" 
                                value={this.state.count}
                                onChange={this.change}/>
                </div>

                <div>
                    {/* Uncontrolled Input */}
                    <input type="number" 
                            placeholder="New Count"
                            ref={(ref) => { this.inpRef = ref}}/>&nbsp;&nbsp;
                    <button onClick={this.setValue}>Set Count</button>
                </div>

                <div>
                    <button id="incBtn" onClick={this.inc}>Increment</button> &nbsp;&nbsp;
                    <button onClick={this.decr}>Decrement</button>
                </div>

            </div>
        );
    }
}

export default Counter;