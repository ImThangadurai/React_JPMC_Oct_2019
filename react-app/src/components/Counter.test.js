import React from 'react'
import Counter from './Counter';

import {mount, shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {create} from 'react-test-renderer';


configure({adapter: new Adapter()})


//Test Suite
describe("CounterTests", () => {

    //Unit Test
    fit("should be created", () => {
        // var x = 10;
        // expect(x).toBe(1);
        const wrapper= mount(<Counter value="0"/>);
        const instance = wrapper.instance();
        expect(instance).toBeTruthy();

    });

    fit("should have state.count = 0", () => {
        // var x = 10;
        // expect(x).toBe(1);
        const wrapper= mount(<Counter value="0"/>);
        const state = wrapper.state();
        const count = state.count;
        expect(count).toBe(0);

    });
    fit("should have state.count incremented by 1", () => {
        // var x = 10;
        // expect(x).toBe(1);
        const wrapper= mount(<Counter value="0"/>);
        const instance = wrapper.instance();
        instance.inc();
        const state = wrapper.state();
        const count = state.count;
        expect(count).toBe(1);

    });

    fit("should have state.count incremented by 1 on button click", () => {
        // var x = 10;
        // expect(x).toBe(1);
        const wrapper= shallow(<Counter value="0"/>);
        
        wrapper.find("#incBtn").simulate('click');
        //wrapper.find("#incBtn").simulate('click');

        const state = wrapper.state();
        const count = state.count;
        expect(count).toBe(1);

    });

    fit("snapshot", () => {
       
        const wrapper= create(<Counter value="0" title="Counter"/>);
        expect(wrapper.toJSON()).toMatchSnapshot();
       

    });


});