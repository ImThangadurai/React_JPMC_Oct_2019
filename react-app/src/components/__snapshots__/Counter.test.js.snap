// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterTests snapshot 1`] = `
<div>
  <h3>
    Counter
  </h3>
  <h5>
    Counter: 
    0
  </h5>
  <div>
    Count: 
    <input
      onChange={[Function]}
      placeholder="Count"
      type="number"
      value={0}
    />
  </div>
  <div>
    <input
      placeholder="New Count"
      type="number"
    />
      
    <button
      onClick={[Function]}
    >
      Set Count
    </button>
  </div>
  <div>
    <button
      id="incBtn"
      onClick={[Function]}
    >
      Increment
    </button>
       
    <button
      onClick={[Function]}
    >
      Decrement
    </button>
  </div>
</div>
`;
